### PR TITLE
GLM: Bot Stuff™

### PR DESCRIPTION
**Pull request by necessarily-equal**
Here is a sizeable chunk from the glm branch. It touches mostly the bot stuff. Changes that aren't pure conversion includes inlining a function and making one class member private with a getter and some friend functions.

To be squashed at merge time

### CODE REVIEW

File: src/sgame/sg_bot_util.h [hunk 1]
```diff
@@ -155,4 +156,10 @@ bool  BotPathIsWalkable( const gentity_t *self, botTarget_t target );
 //consider bot to be stuck if it does not move farther than this in some period of time
 constexpr float BOT_STUCK_RADIUS = 150;
 
+//Those are not really BG, but hopefully someday they will be.
+glm::vec3 BG_GetClientNormal( const playerState_t *ps );
```
Comment (slipher) [inline src/sgame/sg_bot_util.h:160]: Do not write functions prefixed BG_ in sgame.

File: src/sgame/botlib/bot_types.h [hunk 2]
```diff
@@ -60,6 +62,9 @@ struct botNavCmd_t
 	float    dir[ 3 ]; // normalized direction of the target
 	int      directPathToGoal;
 	int      havePath;
+	glm::vec3 glm_dir ( void ) const { return glm::vec3( dir[0] , dir[1] , dir[2] ); };
```
Comment (slipher) [inline src/sgame/botlib/bot_types.h:65]: Do not write `void` in function prototypes.

File: src/sgame/sg_bot_local.h [hunk 3]
```diff
@@ -115,12 +116,19 @@ struct botMemory_t
...
-	vec3_t      futureAim;
-	botNavCmd_t nav;
+	glm::vec3   futureAim;
+private:
+	botNavCmd_t m_nav;
+public:
+	botNavCmd_t const& nav( void ) const { return m_nav; };
+	void clearNav( void ) { memset( &m_nav, 0, sizeof( m_nav ) ); }
+	void setGoal( botTarget_t target, bool direct ) { goal = target; m_nav.directPathToGoal = direct; }
+	friend void G_BotThink( gentity_t * );
```
Comment (slipher) [inline src/sgame/sg_bot_local.h:126]: This friend stuff is very ugly. Maybe if it were going to some sort of sensible end goal but I doubt it...

File: src/sgame/sg_bot_util.h (hunk 1 above)
Comment (necessarily-equal) [reply src/sgame/sg_bot_util.h:160]: This is fixed in a later commit. Not sure if you want me to cherry-pick even more commits in this PR or if you are ok with waiting on the next one.

File: src/sgame/sg_bot_local.h (hunk 3 above)
Comment (necessarily-equal) [reply src/sgame/sg_bot_local.h:126]: I think the idea is to identify the functions which use it and have a easier time to check they are correct. and progressively migrate to correct setters or smth.

It's not my code so I don't really know though

File: src/sgame/botlib/bot_types.h (hunk 2 above)
Comment (necessarily-equal) [reply src/sgame/botlib/bot_types.h]: Fixed

Comment (slipher) [reply src/sgame/botlib/bot_types.h]: There's still some `( void )` left in there

File: src/sgame/sg_bot_util.h (hunk 1 above)
Comment (slipher) [reply src/sgame/sg_bot_util.h:160]: I think it's better to assume that a contributor may vanish at any moment. So no bad code should be merged with the promise of being fixed some time in the future.

Comment (necessarily-equal) [reply src/sgame/sg_bot_util.h:160]: Ok but like, the code is already written.

So would you prefer more commits in this PR or a second one?

File: src/sgame/botlib/bot_types.h (hunk 2 above)
Comment (necessarily-equal) [reply src/sgame/botlib/bot_types.h]: Right, removed the last one. Trusting github to show the whole PR was a bad idea :S

File: src/sgame/sg_bot_util.h (hunk 1 above)
Comment (slipher) [reply src/sgame/sg_bot_util.h:160]: Regardless of how you split it I prefer to have the commit that fixes it in the same one that messes it up.

Comment (necessarily-equal) [reply src/sgame/sg_bot_util.h]: I've added a commit for that (to be squashed with everything else)

File: src/sgame/sg_bot_ai.cpp [hunk 4]
```diff
@@ -1282,8 +1280,8 @@ AINodeStatus_t BotActionRepair( gentity_t *self, AIGenericNode_t *node )
 		G_ForceWeaponChange( self, WP_HBUILD );
 	}
 
-	AngleVectors( self->client->ps.viewangles, forward, nullptr, nullptr );
-	VectorMA( self->s.origin, self->r.maxs[1], forward, selfPos );
```
Comment (slipher) [inline src/sgame/sg_bot_ai.cpp:1287]: Why is this VectorMA deleted?

File: src/sgame/sg_bot_util.cpp [hunk 5]
```diff
@@ -955,11 +955,18 @@ botTarget_t BotGetRetreatTarget( const gentity_t *self )
 botTarget_t BotGetRoamTarget( const gentity_t *self )
 {
 	botTarget_t target;
-	vec3_t targetPos;
+	glm::vec3 point;
 
-	BotFindRandomPoint( self->s.number, targetPos );
+	int botClientNum = self->s.number;
+	if ( !BotFindRandomPointInRadius( botClientNum, g_entities[ botClientNum ].s.origin, point, 2000 ) )
```
Comment (slipher) [inline src/sgame/sg_bot_util.cpp]: Use `self` instead of going through g_entities

File: src/sgame/sg_bot_ai.cpp (hunk 4 above)
Comment (necessarily-equal) [reply src/sgame/sg_bot_ai.cpp:1287]: The result seems to be ignored.

File: src/sgame/sg_bot_util.cpp (hunk 5 above)
Comment (necessarily-equal) [reply src/sgame/sg_bot_util.cpp]: done

**Comment by slipher** [comment]
LGTM

**Comment by DolceTriade** [comment]
yay!!

**Pull request merged**